File: src/2020/day2/password_philosophy.py
from pathlib import Path
from typing import Callable


def is_password_sled_valid(first: int, second: int, letter: str, password: str) -> bool:
    return first <= password.count(letter) <= second


def is_password_tobogan_valid(
    first: int, second: int, letter: str, password: str
) -> bool:
    return (password[first - 1] == letter) ^ (password[second - 1] == letter)


def validate_passwords(
    passwords_path: Path, is_password_valid_fn: Callable[[int, int, str, str], bool]
) -> None:
    with open(passwords_path) as file:
        passwords_w_policy = file.read().splitlines()
    valid_passwords = []
    for password_w_policy in passwords_w_policy:
        first, second, letter, password = (
            password_w_policy.replace("-", " ").replace(":", " ").split()
        )
        if is_password_valid_fn(int(first), int(second), letter, password):
            valid_passwords.append(password_w_policy)
    print(f"{len(valid_passwords)} valid passwords were found: ")
    print("\n".join(valid_passwords))


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(
        description="AoC 2020 - Day 2: Password Philosophy"
    )
    parser.add_argument(
        "passwords_path", type=Path, help="The path to the passwords file to read"
    )
    parser.add_argument(
        "password_policy",
        type=str,
        choices=["sled", "tobogan"],
        help="The policy to use to determine if the password is valid",
    )
    args = parser.parse_args()
    validate_passwords(
        args.passwords_path, locals()[f"is_password_{args.password_policy}_valid"]
    )

File: src/2020/day2/password_philosophy.rs
use std::fs;
use std::path::{Path, PathBuf};

use regex::Regex;
use structopt::StructOpt;

type ValidatePasswordFn = fn(usize, usize, &char, &str) -> bool;

fn is_password_sled_valid(first: usize, second: usize, letter: &char, password: &str) -> bool {
    let letter_count = password.matches(letter).count();
    (first <= letter_count) && (letter_count <= second)
}

fn is_password_tobogan_valid(first: usize, second: usize, letter: &char, password: &str) -> bool {
    let password_chars = password.chars().collect();
    (password_chars[first - 1] == letter) ^ (password_chars[second - 1] == letter)
}

fn validate_passwords(passwords_path: &Path, is_password_valid_fn: ValidatePasswordFn) {
    let passwords_w_policy = fs::read_to_string(passwords_path)
        .expect(format!("No file found at: {:?}.", passwords_path).as_str())
        .lines();
    let mut valid_passwords: Vec<&str> = Vec::new();
    let separator = Regex::new(r"([ -:]+)").expect("Invalid regex");
    for password_w_policy in passwords_w_policy {
        let splits = separator.split(password_w_policy).collect();
        println!("{}", splits);
        valid_passwords.push(&password_w_policy)
    }
    println!("{} valid passwords were found:", valid_passwords.len());
}

/// AoC 2020 - Day 2: Password Philosophy
#[derive(StructOpt)]
struct Cli {
    /// The path to the passwords file to read
    #[structopt(parse(from_os_str))]
    passwords_path: PathBuf,
    /// The policy to use to determine if the password is valid
    password_policy: String,
}

fn main() {
    let args = Cli::from_args();
    if args.password_policy == "sled" {}
    let is_password_valid_fn = match args.password_policy.as_str() {
        "sled" => is_password_sled_valid,
        "tobogan" => is_password_tobogan_valid,
        _ => panic!(
            "The value for input argument `password_policy` is not supported. \
        It should be one of: ['sled', 'tobogan']."
        ),
    };
    validate_passwords(&args.passwords_path, is_password_valid_fn);
}
